### PR TITLE
[#1530] Create new controls UI for video playback

### PR DESCRIPTION
**Pull request by kroucis**
Addresses #1530

Create a new controls UI layer for the DamusVideoPlayer, providing play/pause, volume control, content scale, playback rate controls, as well as a seek/timecode bar.

Screenshot 2023-09-05 at 11 42 46
Screenshot 2023-09-05 at 11 42 55

## Features
- Scrub on the horizontal seek bar to change the playback time
- Play/Pause button
- Mute/Unmute button and volume vertical slider
- Select playback rate from 0.5x up to 2.0x
- Tap video to hide/show controls
- Swap between aspect fill and aspect fit
- Player has option to hide controls on startup
- Player's video tap behavior can be customized
## Fixes
- The VideoPlayer has been reworked from frame-by-frame updating to an event-driven model, significantly improving performance and understandability.
- The VideoPlayer and DamusVideoPlayer had a number of leaks that were preventing resource cleanup and causing a lot of conflicting video playback behavior.
## TODOs
- Animate the volume slider in/out
- Animate the entire control layer in/out
- Create a timer to automatically hide the controls after some time of inactivity
- Color/Theme unification
## Known Issues
- This still suffers from [#1386] as that is a navigation/data model architecture issue, not specific to the video player.
## Notes
- !! This feature *DOES NOT* INCLUDE A FIX FOR #1386 ! That fix will be provided in a future PR. !!
## Support
Bounty: 100k satoshis(BTC) -> 0.001 BTC
Payout Address: `bc1q5sxugecyhq5stxn4d79p4908hdr7ljcn35uyk7`
btc001

### CODE REVIEW

**Comment by kroucis** [comment]
The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote:
>The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...

test suite is very buggy

**Comment by kroucis** [comment]
Well I tested these changes on a 16.0 simulator instance. Compiled and ran just fine.

**Comment by jb55** [comment]
Reviewed-by: William Casarin ***@***.***>

**Comment by bryanmontz** [comment]
Yeah it’ll definitely conflict. I’m removing GSPlayer and making a lot of simplifications in DamusVideoPlayer. If you want, you can finish out this task and merge it when you’re happy, then I can deal with the conflicts. I might be able to contribute a review tonight, but not sure.



On Sep 5, 2023, at 11:49 AM, William Casarin ***@***.***> wrote:


Hey Kyle & Bryan,

On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote:
>Addresses #1530
>
>Create a new controls UI layer for the DamusVideoPlayer, providing play/pause, volume control, content scale, playback rate controls, as well as a seek/timecode bar.
>
>&lt;img width=&quot;594&quot; alt=&quot;Screenshot 2023-09-05 at 11 42 46&quot; src=&quot;https://github.com/damus-io/damus/assets/1031501/de89196e-0951-4a59-8430-99a8cd166e60&quot;&gt;
>&lt;img width=&quot;586&quot; alt=&quot;Screenshot 2023-09-05 at 11 42 55&quot; src=&quot;https://github.com/damus-io/damus/assets/1031501/4a947870-640d-4c7e-89f7-1eab30326aa3&quot;&gt;

Definitely needs a design pass. not sure about the purple. Not a blocker though, we can fix this up after.

>Current features:
>[..]
>- The VideoPlayer has been reworked from frame-by-frame updating to an event-driven model, significantly improving performance and understandability.
>Known Issues:
>- This still suffers from [#1386] as that is a navigation/data model architecture issue, not specific to the video player.
>Note:
>- This feature DOES NOT INCLUDE A FIX FOR #1386 ! That fix will be provided in a future PR.

Does this conflict with the updates you're making Bryan?


Will

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 09:55:55AM -0700, William Casarin wrote:
>On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote:
>>The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...
>
>test suite is very buggy

I would still like to not break the CI though, we should try to
workaround this somehow.

**Comment by kroucis** [comment]
> On Tue, Sep 05, 2023 at 09:55:55AM -0700, William Casarin wrote: On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote: >The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug... test suite is very buggy
> I would still like to not break the CI though, we should try to workaround this somehow.

A simple but unappealing fix would be to NOT use the popover call. This turns the popover into a sheet. Looks like crap, but is supported on all versions of SwiftUI

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 10:18:50AM -0700, Bryan Montz wrote:
>Yeah it’ll definitely conflict. I’m removing GSPlayer and making a lot of simplifications in DamusVideoPlayer. If you want, you can finish out this task and merge it when you’re happy, then I can deal with the conflicts. I might be able to contribute a review tonight, but not sure.

As long as we can still remove GSPlayer after this change then I'm
happy. I prefer not to have dependencies that we don't need.

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 01:42:00PM -0400, Grimless wrote:
>Closes: https://github.com/damus-io/damus/pull/1537
>---
> damus/Views/Video/DamusVideoPlayer.swift | 7 ++-----
> 1 file changed, 2 insertions(+), 5 deletions(-)
>
>diff --git a/damus/Views/Video/DamusVideoPlayer.swift b/damus/Views/Video/DamusVideoPlayer.swift
>index b6e92346f..1ae325347 100644
>--- a/damus/Views/Video/DamusVideoPlayer.swift
>+++ b/damus/Views/Video/DamusVideoPlayer.swift
>@@ -122,11 +122,8 @@ struct DamusVideoPlayer: View {
>             .popover(isPresented: $isOptionsPopoverPresented,
>                      attachmentAnchor: .point(.top),
>                      arrowEdge: .top) {
>-                if #available(iOS 16.4, macOS 13.3, *) {
>-                    OptionsMenu
>-                        .presentationCompactAdaptation(.popover)
>-                }

I think we should just fix the CI instead of dropping this if it looks better.

>-                else {
>+                VStack {
>+                    Text("Playback Speed").font(.title)
>                     OptionsMenu
>                 }
>             }

**Comment by jb55** [comment]
On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote:
>Current features:
>- Scrub on the horizontal seek bar to change the playback time
>- Play/Pause button
>- Mute/Unmute button and volume vertical slider
>- Select playback rate from 0.5x up to 2.0x
>- Tap video to hide/show controls
>- Player has option to hide controls on startup
>- Player&#39;s video tap behavior can be customized

It looks like the fullscreen button doesn't do anything? I don't see it
listed here so I assume that's not implemented?

**Comment by kroucis** [comment]
> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?

No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.

**Comment by jb55** [comment]
I'm hitting crashes, here one I just hit when testing this branch

Fatal error: Attempted to read an unowned reference but the object was already deallocated

Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001d55c401c in __pthread_kill ()
#1	0x00000001f7ded680 in pthread_kill ()
#2	0x0000000197ea3bb0 in abort ()
#3	0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) ()
#4	0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) ()
#5	0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) ()
#6	0x0000000189766018 in swift_unknownObjectUnownedLoadStrong ()
#7	0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345
#8	0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) ()
#9	0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
#10	0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
#11	0x000000010069dc54 in thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
#12	0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()

VideoPlayer.swift:342
342   observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in
343       guard let self else { return }
344       
345       Task { @MainActor in  // <--- !!!CRASH HERE !!!
346           self.videoPlayer.model.currentTime = uiView.currentDuration
347       }
348       
349       self.updateBuffer(uiView: uiView)
350   }

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 11:50:21AM -0700, Kyle Roucis wrote:
>> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
>> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?
>
>No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.

ok we should remove that button then until we have fullscreen.

**Comment by kroucis** [comment]
> On Tue, Sep 05, 2023 at 11:50:21AM -0700, Kyle Roucis wrote:
> >> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
> >> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?
> >
> >No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.
> 
> ok we should remove that button then until we have fullscreen.
> 

Oh! Uh are you sure? I still find the aspect change to be helpful even without fullscreen support.

**Comment by kroucis** [comment]
> 
> I'm hitting crashes, here one I just hit when testing this branch
> 
> Fatal error: Attempted to read an unowned reference but the object was already deallocated
> 
> Thread 1 Queue : com.apple.main-thread (serial)
> #0	0x00000001d55c401c in __pthread_kill ()
> #1	0x00000001f7ded680 in pthread_kill ()
> #2	0x0000000197ea3bb0 in abort ()
> #3	0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) ()
> #4	0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) ()
> #5	0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) ()
> #6	0x0000000189766018 in swift_unknownObjectUnownedLoadStrong ()
> #7	0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345
> #8	0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) ()
> #9	0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
> #10	0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
> #11	0x000000010069dc54 in thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
> #12	0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
> 
> VideoPlayer.swift:342
> 342   observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in
> 343       guard let self else { return }
> 344       
> 345       Task { @MainActor in  // <--- !!!CRASH HERE !!!
> 346           self.videoPlayer.model.currentTime = uiView.currentDuration
> 347       }
> 348       
> 349       self.updateBuffer(uiView: uiView)
> 350   }
> 

Oh! I didn't see any of that. I can change the unowned references to weak references. Use of unowned should be rare anyway. Man, this is a lot of work for $26...

**Comment by jb55** [comment]
On Tue, Sep 05, 2023 at 12:00:54PM -0700, Kyle Roucis wrote:
>> >No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.
>>
>> ok we should remove that button then until we have fullscreen.
>>
>
>Oh! Uh are you sure? I still find the aspect change to be helpful even without fullscreen support.

yes, the button did nothing on all the videos I tried it on.

**Comment by kroucis** [comment]
> I'm hitting crashes, here one I just hit when testing this branch Fatal error: Attempted to read an unowned reference but the object was already deallocated Thread 1 Queue : com.apple.main-thread (serial) #0 0x00000001d55c401c in __pthread_kill () #1 0x00000001f7ded680 in pthread_kill () #2 0x0000000197ea3bb0 in abort () #3 0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) () #4 0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) () #5 0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) () #6 0x0000000189766018 in swift_unknownObjectUnownedLoadStrong () #7 0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345 #8 0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) () #9 0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @sendable @async () -> () () #10 0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @sendable @async () -> () () #11 0x000000010069dc54 in thunk for @escaping @callee_guaranteed @sendable @async () -> ***@***.*** τ_0_0) () #12 0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @sendable @async () -> ***@***.*** τ_0_0) () VideoPlayer.swift:342 342 observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in 343 guard let self else { return } 344 345 Task { @mainactor in // <--- !!!CRASH HERE !!! 346 self.videoPlayer.model.currentTime = uiView.currentDuration 347 } 348 349 self.updateBuffer(uiView: uiView) 350 }

I have removed the fullscreen button and changed the unowned references to weak. I have been unable to recreate the crash you listed, so any more information would be helpful. Also removing the button fails the acceptance criteria for this issue. What is to be done about that?

**Comment by jb55** [comment]
>I have removed the fullscreen button and changed the unowned references to weak. I have been unable to recreate the crash you listed, so any more information would be helpful. Also removing the button fails the acceptance criteria for this issue. What is to be done about that?

Not sure what you mean, isn't the acceptance criteria is to be able to
enter an exit fullscreen? This PR doesn't add fullscreen like you
mentioned?

File: damus/Views/Video/VideoPlayer.swift [hunk 1]
```diff
@@ -166,27 +188,33 @@ public extension VideoPlayer {
     }
 }
 
-func get_video_size(player: AVPlayer) async -> CGSize? {
-    let res = Task.detached(priority: .background) {
-        return player.currentImage?.size
+fileprivate extension AVPlayer {
+    var videoSize: CGSize? {
+        get async {
+            let res = Task.detached(priority: .background) {
```
Comment (bryanmontz) [inline damus/Views/Video/VideoPlayer.swift:194]: It's not a good idea to use `.background` priority for something that you want to update the UI for the user. I would recommend using the default value, which is `.userInitiated`.

**Review by bryanmontz — COMMENTED** [review]
In my opinion, while this is a sorely needed feature add, this PR adds complexity to an already confusing architecture for the video player. I think the addition of UI controls could be done in a way that doesn't interfere so much with the player view, which ideally should just be rendering the video based on the current video settings. There are also lots of non-functional changes (such as changing `mute_icon_color` to `muteIconColor`) mixed in with functional changes, which makes it more difficult to review.

**Comment by jb55** [comment]
@kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!

**Comment by kroucis** [comment]
> @kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!

Yes, I understand, I've been in software for a long time. But it's clear that this approach to the issue is not acceptable, so another attempt will have to be made. Once @bryanmontz 's video player changes have been made and introduced, a simpler UI that is rendering the video based on the current video settings can be implemented. Since the fullscreen criteria cannot be met at this time, the reviewers have decided these changes are too complex and hard to review, and there may be some unowned reference crash issues, I think it's best to learn from this attempt and try again.

**Comment by jb55** [comment]
On Wed, Sep 06, 2023 at 09:06:21AM -0700, Kyle Roucis wrote:
>> @kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!
>
>Yes, I understand, I've been in software for a long time. But it's clear that this approach to the issue is not acceptable, so another attempt will have to be made. Once @bryanmontz 's video player changes have been made and introduced, a simpler UI that is rendering the video based on the current video settings can be implemented. Since the fullscreen criteria cannot be met at this time, the reviewers have decided these changes are too complex and hard to review, and there may be some unowned reference crash issues, I think it's best to learn from this attempt and try again.

ah thanks for clarifying, sounds good and agreed!

**Pull request closed without merging**